document/src/vespa/document/datatype/mapdatatype.cpp

```diff
@@ -55,71 +55,51 @@ MapDataType::operator==(const DataType& other) const
     return (*_keyType == *w->_keyType) && (*_valueType == *w->_valueType);
 }
 
-FieldPath::UP
-MapDataType::buildFieldPathImpl(const DataType &dataType,
+void
+MapDataType::buildFieldPathImpl(FieldPath & path, const DataType &dataType,
                                 const vespalib::stringref &remainFieldName,
-                                const DataType &keyType,
-                                const DataType &valueType)
+                                const DataType &keyType, const DataType &valueType)
 {
     if (!remainFieldName.empty() && remainFieldName[0] == '{') {
         vespalib::string rest = remainFieldName;
         vespalib::string keyValue = FieldPathEntry::parseKey(rest);
 
-        FieldPath::UP path =
-            valueType.buildFieldPath((rest[0] == '.') ? rest.substr(1) : rest);
-        if (!path.get()) {
-            return FieldPath::UP();
-        }
+        valueType.buildFieldPath(path, (rest[0] == '.') ? rest.substr(1) : rest);
 
         if (remainFieldName[1] == '$') {
-            path->insert(path->begin(),
-                         FieldPathEntry(valueType, keyValue.substr(1)));
+            path.insert(path.begin(), FieldPathEntry(valueType, keyValue.substr(1)));
         } else {
             FieldValue::UP fv = keyType.createFieldValue();
             *fv = keyValue;
-            path->insert(path->begin(), FieldPathEntry(valueType, dataType,
-                                 vespalib::CloneablePtr<FieldValue>(fv.release())));
+            path.insert(path.begin(), FieldPathEntry(valueType, dataType, vespalib::CloneablePtr<FieldValue>(fv.release())));
         }
-
-        return path;
     } else if (memcmp(remainFieldName.c_str(), "key", 3) == 0) {
         size_t endPos = 3;
         if (remainFieldName[endPos] == '.') {
             endPos++;
         }
 
-        FieldPath::UP path
-            = keyType.buildFieldPath(remainFieldName.substr(endPos));
-        if (!path.get()) {
-            return FieldPath::UP();
-        }
-        path->insert(path->begin(), FieldPathEntry(dataType, keyType,
-                                                   valueType, true, false));
-        return path;
+        keyType.buildFieldPath(path, remainFieldName.substr(endPos));
+
+        path.insert(path.begin(), FieldPathEntry(dataType, keyType, valueType, true, false));
     } else if (memcmp(remainFieldName.c_str(), "value", 5) == 0) {
```

Comment:
Unrelated to these commits, but isn't this undefined behavior if `remainFieldName` is less than 5 chars long? Also seems like this code will behave strangely if e.g. a field name of "valueSomething" is used, since this code path will be triggered even if that's not the intention.

Same thing applies to handling of "key" above.